src/main/java/com/exactpro/th2/taskutils/FutureTracker.java

```diff
@@ -23,37 +23,138 @@
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 /**
  * Following class tracks futures and when needed tries to wait for them
  */
 public class FutureTracker<T> {
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Condition onRemove = lock.writeLock().newCondition();
+    /** Guarded by {@link #lock} */
     private final Set<CompletableFuture<T>> futures;
     private volatile boolean enabled;
+    private final int limit;
+    private final boolean hasLimit;
 
-    public FutureTracker() {
+    private FutureTracker(int limit) {
         this.futures = new HashSet<>();
         this.enabled = true;
+        this.limit = limit;
+        this.hasLimit = limit > 0;
+    }
+
+    /**
+     * Creates unlimited FutureTracker
+     * @param <T> is trackable {@link CompletableFuture} type
+     * @return instance of FutureTracker class
+     */
+    public static <T> FutureTracker<T> create() {
+        return new FutureTracker<>(0);
+    }
+
+    /**
+     * Creates FutureTracker with limit size
+     * @param <T> is trackable {@link CompletableFuture} type
+     * @param limit is max number of trackable futures if positive value, otherwise tracker unlimited
+     * @return instance of FutureTracker class
+     */
+    public static <T> FutureTracker<T> create(int limit) {
+        return new FutureTracker<>(limit);
     }
 
     /**
      * Track a future
      * @param future to be tracked
+     * @return true if future isn't done and is added for tracking otherwise false
      */
-    public void track(CompletableFuture<T> future) {
+    public boolean track(CompletableFuture<T> future) {
         if (enabled) {
             if (future.isDone()) {
-                return;
+                return false;
             }
-            synchronized (futures) {
+            lock.writeLock().lock();
+            try {
+                if (hasLimit && futures.size() == limit) {
+                    return false;
+                }
```

Comment:
note: this is a behavior-breaking change - previously the method guaranteed that the future was added to the tracking list. I think we should wait here until we have a free space in the `futures` set (like it does the blocking queue when you call `put` method)